### DISCUSSION TITLE
How to hide a part of the transition

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#1058] [Q&A]
Hello,
I am working on transitions and links between elements inside my graph.

I would like to hide the first part of the transition and display for example only the last 2cm from destination element's side.

any suggestions would be helpful, thank you in advance.

**Comment by rsoika** [comment]
Hi @Adem-Kri 
you can overwrite the default view for an edge and register the new view in your DiagramModule

` configureModelElement(context, 'sequenceFlow', BPMNEdge, BPMNEdgeView);`

We have done this in our own BPMN Diagram to add 'rounded' corners. The `PolylineEdgeViewWithGapsOnIntersections` is a good base class for custom edge views.

You can take a look at our implementation [here](https://github.com/imixs/open-bpmn/blob/master/open-bpmn.glsp-client/open-bpmn-glsp/src/bpmn-routing-views.tsx).

We do the rounded corners and also a BPMN specific symbol on the start of a view

![image](https://github.com/eclipse-glsp/glsp/assets/1339124/79e2294c-b561-485e-92e5-6c5dbe21c8fc)

So everything is possible.

**Comment by Adem-Kri** [comment]
@rsoika, I'm not sure if I explained my issue clearly or if I misunderstood your previous response. I tried your suggestion but couldn't solve my problem. Can you explain it more clearly?

this is my requirement: for a link, I have an option where I can hide a part of the transition as I told you the last part of the link

![222](https://github.com/eclipse-glsp/glsp/assets/116633370/d8265d1b-8c9d-485a-bc19-9b9e7b41efaf)


![111](https://github.com/eclipse-glsp/glsp/assets/116633370/377097a2-7937-47f0-ac04-40533166353b)

**Reply by rsoika** [reply]
Yes but you can do this also when you set a CSS rule as transparent for the edge and only set the class="sprotty-edge arrow" with a color.
Or you draw the complete line in a custom viewer as We have done it in our example code.

**Comment by CamilleLetavernier** [comment]
Hi,

The edge-line (and, optionally, the arrow decoration) is rendered by the EdgeView configured for your element. For example, a `PolylineEdgeView`.

The idea is that you can change this view (use a different, custom implementation), where you override the "renderLine" method. For example, instead of rendering all segments of the line, you can decide to render only the last segment, or even just a small portion of the last segment.

**Reply by Adem-Kri** [reply]
Hello @CamilleLetavernier, thank you for the quick and precise response.

I'm still a little bit confused so i'll give you my edgeView code so that maybe you could help me do the implementation of your solution

```ts
@injectable()
export class EdgeView extends PolylineEdgeViewWithGapsOnIntersections {
  protected override renderAdditionals(
    edge: SEdge,
    segments: Point[],
    context: RenderingContext
  ): VNode[] {
    const additionals = super.renderAdditionals(edge, segments, context);
    const p1 = segments[segments.length - 2];
    const p2 = segments[segments.length - 1];
    const firstPoint = segments[0];
    const lastPoint = segments[segments.length-1];
    
    const headArrow: any = (
      <path
        d="M 1,0 L 10,-4 L 10,4 Z"
        transform={`rotate(${toDegrees(
          angleOfPoint({ x: p1.x - p2.x, y: p1.y - p2.y })
        )} ${p2.x} ${p2.y}) translate(${p2.x} ${p2.y})`}
      />
    );

    const rootArrow: any = <circle cx={firstPoint.x} cy={firstPoint.y} r={5} />;

    additionals.push(rootArrow);
    additionals.push(headArrow);
    // additionals.push(midlleLabelNode);
    return additionals;
  }
}
```
thanks in advance

**Reply by CamilleLetavernier** [reply]
There are two parts to rendering an edge:

- renderLine(): the actual routed line, from source to target, which follows the defined bendpoints (if any)
- renderAdditionals(): to add extra decorations on top of the line (Typically, an arrow on the source or target connection point)

In your case, you want to hide the main path (from renderLine()). This is the part you need to customize, to remove unnecessary parts.

The default line implementation (From GLSP's PolylineEdgeView) looks like this:

```ts
    protected renderLine(edge: SEdge, segments: Point[], context: RenderingContext, args?: IViewArgs): VNode {
        const firstPoint = segments[0];
        let path = `M ${firstPoint.x},${firstPoint.y}`;
        for (let i = 1; i < segments.length; i++) {
            const p = segments[i];
            path += ` L ${p.x},${p.y}`;
        }
        return <path d={path} />;
    }
```

It is an SVG Path that starts from the source anchor point, follows all bendpoints defined for this edge, and ends at the target anchor point.

Since you only care about the last segment, instead of iterating over all individual points (segments), you can render only the last segment (Which is the last 2 points in the "segments" array).

Then, if you want only a subset of this last segment (to only draw the last 2cm), you'll need to do a bit of geometry math to truncate it.

In your case, I think you can ignore the renderAdditionals() method entirely, unless you want to add arrows on this small segment. But I would recommend ignoring this part in the first step.
